ledger-core/v2/virtual/handlers/vdelegatedcallrequest.go

```diff
@@ -0,0 +1,204 @@
+	s := sm.(*SMVDelegatedCallRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+	injector.MustInject(&s.objectCatalog)
+	injector.MustInject(&s.jetCoordinator)
+}
+
+func (*dSMVDelegatedCallRequest) GetInitStateFor(sm smachine.StateMachine) smachine.InitFunc {
+	s := sm.(*SMVDelegatedCallRequest)
+	return s.Init
+}
+
+/* -------- Instance ------------- */
+
+func (s *SMVDelegatedCallRequest) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return dSMVDelegatedCallRequestInstance
+}
+
+func (s *SMVDelegatedCallRequest) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepProcess)
+}
+
+func (s *SMVDelegatedCallRequest) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	objectRef := s.Payload.Callee
+	s.objectSharedState = s.objectCatalog.GetOrCreate(ctx, objectRef)
+
+	return ctx.Jump(s.stepWaitObjectReady)
+}
+
+func (s *SMVDelegatedCallRequest) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	var (
+		semaphoreReadyToWork smachine.SyncLink
+	)
+
+	action := func(state *object.SharedState) {
+		semaphoreReadyToWork = state.ReadyToWork
+	}
+
+	switch s.objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
+	case smachine.Passed:
+	case smachine.NotPassed:
+		return ctx.WaitShared(s.objectSharedState.SharedDataLink).ThenRepeat()
+	default:
+		panic(throw.Impossible())
+	}
+
+	if ctx.AcquireForThisStep(semaphoreReadyToWork).IsNotPassed() {
+		return ctx.Sleep().ThenRepeat()
+	}
+
+	return ctx.Jump(s.processRequest)
+}
+
+func (s *SMVDelegatedCallRequest) processRequest(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	var (
+		tooOldRequest bool
+
+		pendingTableFull bool
+	)
+
+	action := func(state *object.SharedState) {
+		var (
+			oldestPulse  pulse.Number
+			pendingList  *object.RequestList
+			pendingCount uint8
+		)
+
+		callTolerance := s.Payload.CallFlags.GetInterference()
+
+		switch callTolerance {
+		case contract.CallTolerable:
+			pendingList = state.PendingTable.GetList(contract.CallTolerable)
+			oldestPulse = state.OrderedPendingEarliestPulse
+			pendingCount = state.ActiveOrderedPendingCount
+		case contract.CallIntolerable:
+			pendingList = state.PendingTable.GetList(contract.CallIntolerable)
+			oldestPulse = state.UnorderedPendingEarliestPulse
+			pendingCount = state.ActiveUnorderedPendingCount
+		default:
+			panic(throw.Unsupported())
+		}
+
+		if oldestPulse == pulse.Unknown || s.Payload.RequestReference.GetLocal().GetPulseNumber() < oldestPulse {
+			tooOldRequest = true
+			return
+		}
+
+		// pendingList already full
+		if pendingList.Count() == int(pendingCount) && !pendingList.Exist(s.Payload.RequestReference) {
+			pendingTableFull = true
+			return
+		}
+
+		if !pendingList.Add(s.Payload.RequestReference) {
+			ctx.Log().Warn("request already processed")
```

Comment:
@sexyhamster Maybe add some related info in log message?

Reply:
added more descriptive error